src/util/pg.py

```diff
@@ -117,23 +104,69 @@ def execute(query):
 
         cr.commit()
 
+        CONCURRENCY_ERRORCODES = {
+            errorcodes.DEADLOCK_DETECTED,
+            errorcodes.SERIALIZATION_FAILURE,
+        }
+        failed_queries = []
+        tot_cnt = 0
         with ThreadPoolExecutor(max_workers=max_workers) as executor:
-            return sum(
-                log_progress(
-                    executor.map(execute, queries),
-                    logger,
-                    qualifier="queries",
-                    size=len(queries),
-                    estimate=False,
-                    log_hundred_percent=True,
-                )
-            )
+            future_queries = {executor.submit(execute, q): q for q in queries}
+            for future in log_progress(
+                concurrent.futures.as_completed(future_queries),
+                logger,
+                qualifier="queries",
+                size=len(queries),
+                estimate=False,
+                log_hundred_percent=True,
+            ):
+                try:
+                    tot_cnt += future.result() or 0
+                except psycopg2.OperationalError as exc:
+                    if exc.pgcode not in CONCURRENCY_ERRORCODES:
+                        raise
+
+                    # to be retried without concurrency
+                    failed_queries.append(future_queries[future])
+
+        if failed_queries:
+            logger.info("Serialize queries that failed due to concurrency issues")
+            tot_cnt += _parallel_execute_serial(cr, failed_queries, logger=logger)
+            cr.commit()
+
+        return tot_cnt
 
 else:
     _parallel_execute_threaded = _parallel_execute_serial
 
 
 def parallel_execute(cr, queries, logger=_logger):
+    """
+    Execute queries in parallel, using a maximum of 8 workers.
+
+    The actual number of workers is computed as the minimum between the number of queries
+    and that of the availablue cpu cores, capped at 8.
+
+    .. example::
+        .. code-block:: python
+            util.parallel_execute(cr, ["REINDEX TABLE {}".format(t) for t in tables])
```

Comment:
At least use the correct way of formatting a query.
```suggestion
            util.parallel_execute(cr, [util.format_query(cr, "REINDEX TABLE {}", t) for t in tables])
```